mava/networks/retention.py

```diff
@@ -0,0 +1,309 @@
+        )
+        self.w_k = self.param(
+            "w_k",
+            nn.initializers.normal(stddev=1 / self.embed_dim),
+            (self.embed_dim, self.head_size),
+        )
+        self.w_v = self.param(
+            "w_v",
+            nn.initializers.normal(stddev=1 / self.embed_dim),
+            (self.embed_dim, self.head_size),
+        )
+
+    def __call__(
+        self, key: Array, query: Array, value: Array, hstate: Array, dones: Array
+    ) -> Tuple[Array, Array]:
+        """Chunkwise (default) representation of the retention mechanism."""
+        batch, chunk_size, _ = value.shape
+
+        # Apply projection to q_proj, k_proj, v_proj
+        q_proj = query @ self.w_q
+        k_proj = key @ self.w_k
+        v_proj = value @ self.w_v
+        k_proj = k_proj.transpose(0, -1, -2)
+
+        # Compute next hidden state
+        if self.memory_config.type == "ff_sable":
+            # No decay matrix or xi for FF Sable since we don't have temporal dependencies.
+            decay_matrix = jnp.ones((batch, chunk_size, chunk_size))
+            xi = jnp.ones((batch, chunk_size, 1))
+            next_hstate = (k_proj @ v_proj) + hstate
+        else:
+            decay_matrix = self.get_decay_matrix(dones)
+            xi = self.get_xi(dones)
+            chunk_decay = self.decay_kappa ** (chunk_size // self.n_agents)
+            delta = ~jnp.any(dones[:, :: self.n_agents], axis=1)[:, jnp.newaxis, jnp.newaxis]
+            next_hstate = (
+                k_proj @ (v_proj * decay_matrix[:, -1].reshape((batch, chunk_size, 1)))
+            ) + hstate * chunk_decay * delta
+
+        # Compute the inner chunk and cross chunk
+        cross_chunk = (q_proj @ hstate) * xi
+        inner_chunk = ((q_proj @ k_proj) * decay_matrix) @ v_proj
+
+        # Compute the final retention
+        ret = inner_chunk + cross_chunk
+        return ret, next_hstate
+
+    def recurrent(
+        self, key_n: Array, query_n: Array, value_n: Array, hstate: Array
+    ) -> Tuple[Array, Array]:
+        """Recurrent representation of the retention mechanism."""
+        # Apply projection to q_proj, k_proj, v_proj
+        q_proj = query_n @ self.w_q
+        k_proj = key_n @ self.w_k
+        v_proj = value_n @ self.w_v
+
+        # Apply the retention mechanism and update the hidden state
+        updated_hstate = hstate + (k_proj.transpose(0, -1, -2) @ v_proj)
+        ret = q_proj @ updated_hstate
+
+        return ret, updated_hstate
+
+    def get_decay_matrix(self, dones: Array) -> Array:
+        """Get the decay matrix for the full sequence based on the dones and retention type."""
+        # Extract done information at the timestep level
+        timestep_dones = dones[:, :: self.n_agents]
+
+        # Compute the decay matrix and apply timestep-based masking
+        decay_matrix = self._get_default_decay_matrix(
+            timestep_dones
+        ) * self._get_decay_matrix_mask_timestep(timestep_dones)
```

Comment:
nit + adding shapes would be super useful here (I think these shapes are correct):
```suggestion
        # B, T, T
        timestep_mask = self._get_decay_matrix_mask_timestep(timestep_dones)
        decay_matrix = self._get_default_decay_matrix(timestep_dones)
        decay_matrix *= timestep_mask
```

Reply:
Bump, can we switch to storing the intermediate variables :pray:

Reply:
Ah i forgot this one